### DISCUSSION TITLE
A ListboxOption must have a value prop.

### DISCUSSION THREAD

**Discussion by cristianpascottini** [reach/reach-ui#885] [Q&A]
Hi!
Can you please explain the ratio behind [this](https://github.com/reach/reach-ui/blob/f382d62d9c95e5c249a141106b06bbb76847f4e2/packages/listbox/src/index.tsx#L937-L939)?
```javascript
  if (__DEV__ && !value) {
    throw Error(`A ListboxOption must have a value prop.`);
  }
```
Why is it not possible to let a `ListboxOption` have an empty string as a value? Or even the number`0`?
Are there other ways to do it?

Many thanks.

**Comment by indiesquidge** [comment]
> Why is it not possible to let a ListboxOption have an empty string as a value? Or even the number0?

@cristianpascottini if the Listbox behaved as though it were using React's typings for an [`option` element][1], then `value` should indeed be optional or be allowed to be a number. However, in Reach's Listbox, not having a value for an option would be a bit awkward. (Even in native HTML, the [`value` attribute][2] would need to be required if HTML didn't fall back on a sophisticated approach of finding the nested [text IDL attribute][3] of the option otherwise.)

As to why e.g. a number is not allowed, I suppose a case could be made to allow it, and change the conditional to `if (__DEV__ && value == undefined)`. However, you could implement this yourself at the product level by coercing your values into numbers when they are selected if you really wanted.

```jsx
<Listbox onChange={(value) => {
  console.log(Number(value));
}}>
  <ListboxOption value="0">Option 0</ListboxOption>
  <ListboxOption value="1">Option 1</ListboxOption>
  <ListboxOption value="2">Option 2</ListboxOption>
</Listbox>
```

One of Reach's [philosophies][4] is to be as close to HTML behavior as possible. And while HTML _will_ let you pass e.g. a number as an option's value, the value being submitted as form data will be serialized as a string anyway.

Is there a specific non-string value case you are attempting to create?

[1]: https://github.com/DefinitelyTyped/DefinitelyTyped/blob/afd309b4193c1f448386bf8fe09e512e4422e69e/types/react/index.d.ts#L2342
[2]: https://html.spec.whatwg.org/multipage/form-elements.html#attr-option-value
[3]: https://html.spec.whatwg.org/multipage/form-elements.html#dom-option-text
[4]: https://gist.github.com/ryanflorence/e5c794e6093d16a69fa88d2112a292f7#one-to-one-rendered-dom-element-to-reach-component
